host/src/conversion/limits.rs

```diff
@@ -0,0 +1,426 @@
+//! Limiter for conversion of untrusted data.
+use std::{cell::RefCell, rc::Rc};
+
+use datafusion_common::{DataFusionError, error::Result as DataFusionResult};
+
+/// Limits that should be applied during conversion from untrusted to trusted data.
+#[derive(Debug, Clone)]
+#[expect(missing_copy_implementations, reason = "allow later extensions")]
+pub struct TrustedDataLimits {
+    /// Maximum length of identifiers like names, in bytes.
+    ///
+    /// Also see [`max_aux_string_length`](Self::max_aux_string_length).
+    pub max_identifier_length: usize,
+
+    /// Maximum length of auxiliary strings like error messages or metadata, in bytes.
+    ///
+    /// Does NOT apply to identifiers (see [`max_identifier_length`](Self::max_identifier_length)) or payload data
+    /// like string arrays.
+    pub max_aux_string_length: usize,
+
+    /// Maximum data structure depth.
+    ///
+    /// # Example
+    /// The following tree has depth of 3:
+    ///
+    /// ```text
+    ///    o
+    ///    |
+    /// +--+--+
+    /// |     |
+    /// o     o
+    /// |     |
+    /// |   +-+-+-+
+    /// |   |   | |
+    /// o   o   o o
+    /// ```
+    pub max_depth: u64,
+
+    /// Maximum data structure complexity.
```

Comment:
I think this should really describe how "complexity" is calculated. Is it just the number of items?